### PR TITLE
Adding probability distribution options

### LINKED ISSUE #237: Data types/prior distributions in FlowNet
In FlowNet the only available data types or distributions as of now are uniform and loguniform. ERT offers a few more. We should maybe aim at making it possible to define the same distributions in FlowNet. This means adding (at least) normal, lognormal, triangular, const, and maybe truncated_normal. The remaining ones (errf, derrf and dunif) are maybe a bit more obscure.

Adding more distributions will make some complications for the kriging/constraining to logs - other than that it should not be a problem.

### PR DESCRIPTION
**Pull request by olelod**
*Insert a description of your pull request (PR) here, and check off the boxes below when they are done.*

---

### Contributor checklist

- [x] :tada: This PR closes #237 and #210 
- [x] :scroll: I have broken down my PR into the following tasks:
   - [x] Add possibility of defining more data types/priors in the config yml
        - [x] Normal distribution
        - [x] Truncated normal distribution
        - [x] Lognormal distribution
        - [x] Triangular distribution
        - [x] Dirac distribution (constant)
   - [x] Make all data types available for
        - [x] Bulk volumne
        - [x] Porosity
        - [x] Permeability
        - [x] Aquifer size
        - [x] Fault multipliers
        - [x] Contacts
        - [x] Relative permeability
   - [ ] Update how min/max for a given mean/max or mean/min is found for a loguniform distribution
   - [x] Update the update_distribution function in _run_ahm.py
   - [x] Adding checks in the config parsing to make sure the defined distributions are ok
- [x] :robot: I have added tests, or extended existing tests, to cover any new features or bugs fixed in this PR.
- [x] :book: I have considered adding a new entry in `CHANGELOG.md`.
- [x] :books: I have considered updating the documentation.

### CODE REVIEW

File: src/flownet/parameters/probability_distributions.py [hunk 1]
```diff
@@ -3,39 +3,413 @@
...
+        self.update_distribution(minimum=minimum, mean=mean, maximum=maximum)
+
     @property
     def ert_gen_kw(self) -> str:
-        """string representing an ERT "UNFIORM MIN MAX" distribution keyword for use in GEN_KW"""
-        return f"UNIFORM {self.minimum} {self.maximum}"
+        """string representing an ERT "UNIFORM MIN MAX" distribution keyword for use in GEN_KW"""
+        return f"{self.name} {self.minimum} {self.maximum}"
+
+    def update_distribution(
```
Comment (wouterjdb) [inline src/flownet/parameters/probability_distributions.py:48]: Type annotation

File: src/flownet/parameters/probability_distributions.py [hunk 2]
```diff
@@ -3,39 +3,413 @@
...
+        super().__init__("LOGUNIF")
+        self.update_distribution(minimum=minimum, mean=mean, maximum=maximum)
+
     @property
     def ert_gen_kw(self) -> str:
         """string representing an ERT "LOGUNIF MIN MAX" distribution keyword for use in GEN_KW"""
-        return f"LOGUNIF {self.minimum} {self.maximum}"
+        return f"{self.name} {self.minimum} {self.maximum}"
+
+    def update_distribution(
```
Comment (wouterjdb) [inline src/flownet/parameters/probability_distributions.py:108]: Type annotation

File: src/flownet/parameters/probability_distributions.py [hunk 3]
```diff
@@ -3,39 +3,413 @@
...
+        self.stddev = np.sqrt(
+            (
+                np.log(self.maximum / self.minimum)
+                * (np.power(self.maximum, 2) - np.power(self.minimum, 2))
+                - 2 * np.power(self.maximum - self.minimum, 2)
+            )
+            / (2 * np.power(np.log(self.maximum / self.minimum), 2))
+        )
+
+    def _find_dist_minmax(
```
Comment (wouterjdb) [inline src/flownet/parameters/probability_distributions.py:165]: Type annotation

File: src/flownet/parameters/probability_distributions.py [hunk 4]
```diff
@@ -3,39 +3,413 @@
...
+        super().__init__("TRIANGULAR")
+        self.update_distribution(minimum=minimum, maximum=maximum, mean=mean, mode=mode)
 
     @property
     def ert_gen_kw(self) -> str:
         """string representing an ERT "TRIANGULAR MIN MAX" distribution keyword for use in GEN_KW"""
-        return f"TRIANGULAR {self.minimum} {self.mode} {self.maximum}"
+        return f"{self.name} {self.minimum} {self.mode} {self.maximum}"
+
+    def update_distribution(
```
Comment (wouterjdb) [inline src/flownet/parameters/probability_distributions.py:219]: Type annotation

File: src/flownet/parameters/probability_distributions.py [hunk 5]
```diff
@@ -3,39 +3,413 @@
...
+    def __init__(self, mean, stddev):
+        super().__init__("NORMAL")
+        self.update_distribution(mean=mean, stddev=stddev)
+
+    @property
+    def ert_gen_kw(self) -> str:
+        """string representing an ERT "NORMAL MEAN STDDEV" distribution keyword for use in GEN_KW"""
+        return f"{self.name} {self.mean} {self.stddev}"
+
+    def update_distribution(
```
Comment (wouterjdb) [inline src/flownet/parameters/probability_distributions.py:299]: Type annotation

File: src/flownet/parameters/probability_distributions.py [hunk 6]
```diff
@@ -3,39 +3,413 @@
...
+        self.update_distribution(
+            minimum=minimum, maximum=maximum, mean=mean, stddev=stddev
+        )
+
+    @property
+    def ert_gen_kw(self) -> str:
+        """string representing an ERT "TRUNCATED_NORMAL MEAN STDDEV" distribution keyword for use in GEN_KW"""
+        return f"{self.name} {self.mean} {self.stddev} {self.minimum} {self.maximum}"
+
+    def update_distribution(
```
Comment (wouterjdb) [inline src/flownet/parameters/probability_distributions.py:338]: Type annotation

File: src/flownet/parameters/probability_distributions.py [hunk 7]
```diff
@@ -3,39 +3,413 @@
...
+    def __init__(self, mean, stddev):
+        super().__init__("LOGNORMAL")
+        self.update_distribution(mean=mean, stddev=stddev)
+
+    @property
+    def ert_gen_kw(self) -> str:
+        """string representing an ERT "LOGNORMAL MEAN STDDEV" distribution keyword for use in GEN_KW"""
+        return f"{self.name} {self.mean} {self.stddev}"
+
+    def update_distribution(self, **kwargs):
```
Comment (wouterjdb) [inline src/flownet/parameters/probability_distributions.py:373]: Type annotation

File: src/flownet/parameters/probability_distributions.py [hunk 8]
```diff
@@ -3,39 +3,413 @@
...
+    def __init__(self, constant):
+        super().__init__("CONST")
+        self.update_distribution(constant=constant)
+
+    @property
+    def ert_gen_kw(self) -> str:
+        """string representing an ERT "CONST CONSTANT" distribution keyword for use in GEN_KW"""
+        return f"{self.name} {self.mean}"
+
+    def update_distribution(self, **kwargs):
```
Comment (wouterjdb) [inline src/flownet/parameters/probability_distributions.py:402]: Type annotation

File: src/flownet/parameters/probability_distributions.py [hunk 9]
```diff
@@ -3,39 +3,413 @@
...
+        """
+
```
Comment (wouterjdb) [inline src/flownet/parameters/probability_distributions.py:404]: Docstring

**Review by wouterjdb — CHANGES_REQUESTED** [review]
Wow! This has become a really large PR. I _think_ it is pretty impressive 👍 

I've tried to do my best to go through it but it is nearly impossible to actually still get the complete overview of what the changes are - which means that I cannot really get much further than picking up styling/documentation issues. Checking the actual functionality has become really hard. For the next time, please split your work into smaller tasks such that the reviewer has a better chance of picking up potential issues in the functionality.

Could we maybe have a quick call such that you help me through the code? 🤙

File: src/flownet/ahm/_run_ahm.py [hunk 10]
```diff
@@ -677,19 +682,75 @@ def run_flownet_history_matching(
...
+                None
+                if equil_config_eqlnum[idx].owc_depth is None
+                else equil_config_eqlnum[idx].owc_depth.distribution,
+                None
+                if equil_config_eqlnum[idx].gwc_depth is None
+                else equil_config_eqlnum[idx].gwc_depth.distribution,
+                None
+                if equil_config_eqlnum[idx].goc_depth is None
+                else equil_config_eqlnum[idx].goc_depth.distribution,
+            ],
```
[the remark this answers is not in the document]
Comment (olelod) [reply src/flownet/ahm/_run_ahm.py]: Yes! Also did similiar things a few other places in the code above and below.

File: src/flownet/ahm/_run_ahm.py [hunk 11]
```diff
@@ -719,16 +778,19 @@ def run_flownet_history_matching(
...
-            raise ValueError(
-                f"The aquifer scheme "
-                f"'{aquifer_config['scheme']}' is not valid.\n"
-                f"Valid options are 'global' or 'individual'."
-            )
 
         # Create a pandas dataframe with all parameter definition for each individual tube
         aquifer_dist_values = pd.DataFrame(
-            columns=["parameter", "minimum", "maximum", "loguniform", "aquid"]
+            columns=[
```
[the remark this answers is not in the document]
Comment (olelod) [reply src/flownet/ahm/_run_ahm.py]: made a `column_names_probdist` to use several places

File: src/flownet/config_parser/_config_parser.py [hunk 12]
```diff
@@ -1393,3 +1735,249 @@ def parse_config(
...
+            or {"min", "base", "mean"}.issubset(defined_parameters)
+            or {"max", "base", "mean"}.issubset(defined_parameters)
+        ):
+            raise ValueError(
+                "The triangular distribution needs 'min', 'base', and 'max' to be defined. "
+                "Alternatively two of the former in addition to the 'mean'."
+            )
+
+
+def _check_defined(path_in_config_dict: dict, parameter: str):
```
Comment (wouterjdb) [inline src/flownet/config_parser/_config_parser.py:1970]: Missing type annotation for return value

File: src/flownet/config_parser/_config_parser.py [hunk 13]
```diff
@@ -1393,3 +1735,249 @@ def parse_config(
...
+    """
+
+    Args:
+        path_in_config_dict (str): a location in the config schema dictionary
+        parameter (str): a parameter/dictionary found at the given location
+
+    Returns:
+        Nothing, raises ValueErrors if something is wrong
```
Comment (wouterjdb) [inline src/flownet/config_parser/_config_parser.py:1978]: Incorrect, it does return something and does not throw a ValueError

File: src/flownet/parameters/probability_distributions.py [hunk 14]
```diff
@@ -3,39 +3,615 @@
 """
 
 import abc
+from typing import Optional
+import numpy as np
```
Comment (wouterjdb) [inline src/flownet/parameters/probability_distributions.py:7]:
```suggestion

import numpy as np
```

File: src/flownet/parameters/probability_distributions.py [hunk 15]
```diff
@@ -3,39 +3,615 @@
...
+        * The minimum value of the uniform distribution
+        * The mean value of the uniform distribution
+        * The maximum value of the uniform distribution
+
+    Args:
+        minimum (float): The minimum value of the distribution
+        mean (float): The mean value of the distribution
+        mode (float): The mode of the distribution
+        maximum (float): The maximum value of the distribution
+        stddev (float): The standard deviation of the distribution
```
Comment (wouterjdb) [inline src/flownet/parameters/probability_distributions.py:52]: Is this correct?

**Review by wouterjdb — COMMENTED** [review]
LGTM 👍 (but too large too see the total picture, so I hope you did were very good with testing things!). Some tiny comments left and please change back the CI/CD testdata path.

File: src/flownet/config_parser/_config_parser.py (hunk 12 above)
Comment (olelod) [reply src/flownet/config_parser/_config_parser.py:1970]: Added

File: src/flownet/config_parser/_config_parser.py (hunk 13 above)
Comment (olelod) [reply src/flownet/config_parser/_config_parser.py]: Copy/paste always dangerous....

**Review by wouterjdb — APPROVED** [review]
LGTM 🤞

**Comment by olelod** [comment]
Yes 🤞

**Pull request merged**